src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp

```diff
@@ -0,0 +1,742 @@
+                &volume_mesh](const auto dt_variables_ptr) noexcept {
+            evolution::dg::lift_boundary_terms_gauss_points(
+                dt_variables_ptr, volume_det_inv_jacobian, volume_mesh,
+                direction, boundary_corrections_on_face,
+                magnitude_of_interior_face_normal, face_det_jacobian);
+          });
+    }
+  }
+  // Add TimeDerivative correction to volume time derivatives.
+  if constexpr (uses_time_derivative_condition) {
+    if (volume_mesh.quadrature(0) == Spectral::Quadrature::GaussLobatto) {
+      db::mutate<dt_variables_tag>(
+          box, [&direction, &dt_time_derivative_correction,
+                &volume_mesh](const auto dt_variables_ptr) noexcept {
+            add_slice_to_data(
+                dt_variables_ptr, dt_time_derivative_correction,
+                volume_mesh.extents(), direction.dimension(),
+                index_to_slice_at(volume_mesh.extents(), direction));
+          });
+    } else {
+      db::mutate<dt_variables_tag>(
+          box, [&direction, &dt_time_derivative_correction,
+                &volume_mesh](const auto dt_variables_ptr) noexcept {
+            interpolate_dt_terms_gauss_points(dt_variables_ptr, volume_mesh,
+                                              direction,
+                                              dt_time_derivative_correction);
+          });
+    }
+  }
+}
+
+/*!
+ * \brief Applies the chosen boundary condition using the `boundary_correction`
+ * on all external faces.
+ *
+ * A `tmpl::for_each` loop along with a `typeid` comparison checks which of the
+ * known boundary conditions is being used. Since each direction can have a
+ * different boundary condition, we must check each the boundary condition in
+ * each external direction.
+ */
+template <typename System, size_t Dim, typename DbTagsList,
+          typename BoundaryCorrection>
+void apply_boundary_conditions_on_all_external_faces(
+    const gsl::not_null<db::DataBox<DbTagsList>*> box,
+    const BoundaryCorrection& boundary_correction,
+    const Variables<
+        typename System::compute_volume_time_derivative_terms::temporary_tags>&
+        temporaries,
+    const Variables<
+        db::wrap_tags_in<::Tags::Flux, typename System::flux_variables,
+                         tmpl::size_t<Dim>, Frame::Inertial>>& volume_fluxes,
+    const Variables<
+        db::wrap_tags_in<::Tags::deriv, typename System::gradient_variables,
+                         tmpl::size_t<Dim>, Frame::Inertial>>& partial_derivs,
+    const Variables<detail::get_primitive_vars_tags_from_system<System>>* const
+        primitive_vars) noexcept {
+  using derived_boundary_conditions = tmpl::remove_if<
+      typename System::boundary_conditions_base::creatable_classes,
+      tmpl::or_<
+          std::is_base_of<domain::BoundaryConditions::MarkAsPeriodic, tmpl::_1>,
+          std::is_base_of<domain::BoundaryConditions::detail::MarkAsNone,
```

Comment:
1. I'm a :banana: Yes, should be cleaned up
2. Sorry, I don't fully understand what you mean. So the current design is that each system can specify whether it wants to support periodic or none boundary conditions by adding them to the `creatable_classes`. If you're saying "add those to the end" then I'd have 2 worries: 1. it doesn't really reduce the type manipulation here since some systems may not support periodic or none, so still need to check 2. having a special location in the typelist is something that users will inevitably get wrong (I will probably get it wrong...)
3. Right, so the reason outflow is an actual boundary condition is because the Outflow conditions should check that the char speeds are all pointing out of the domain. This allows the code to `ERROR` if the outflow condition is violated. This is needed at the excision surface. This allows us to avoid the lifting-type math, but still check the signs of the speeds. Please let me know where (could be multiple places) this should be documented :)

Reply:
1. 👍 
2. I meant the System could have two typelists: a smaller one with the BCs that need to be applied, and a bigger one containing the BCs to apply + (optionally) None + (optionally) Periodic. Benefit: you reduce typelist manipulations like having to remove None and Periodic here; cost: you have make the programmer do the manipulation by specifying two nearly-identical typelists in the System. I'm not saying it's a good idea, but it's an idea. What do you think?
3. I think the code is OK as is... I was just checking my understanding, and it seems I was on the right page. I see further down you have an if block handling the Outflow BCs and that is well documented, so I am satisfied 👍

Reply:
2. I think it's reasonable, my worry is that this leaks the details of how boundary conditions are applied into the user interface, for what is likely minimal to no benefit. The typelist manipulation here is going to be basically instantaneous to compile until we get to several hundred boundary conditions. I favor the current design because it means that users can list all boundary conditions in one place without caring about any of the details as to how they're actually handled :shrug:
3. :+1:

Reply:
2. Sounds good, I agree with your reasoning 👍